c10/core/TensorImpl.h

```diff
@@ -1438,6 +1438,7 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
   }
 
   bool is_non_overlapping_and_dense() const {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(support_as_strided());
```

Comment:
If you're serious about this invariant, it (1) needs docs, and (2) it is better to test for the invariant at sites which modify `is_non_overlapping_and_dense` or `support_as_strided`, which will reduce the number of redundant tests (no point in rechecking the invariant on every access to the tensor, when the only time the invariant could be violated is on writes)

Reply:
@ezyang I took another look at TensorImpl and I'm not so sure about this invariant now. All of following fields seem to be dependent on assumption that support as_strided, if we want to enforce the invariant we need to do it for all of them. (is this list growing? ) 
Plus technically we only need to set support_as_strided once, while all the following fields can change with tensors, now I'm thinking we should use support_as_strided as bitfield and let the following fields condition on it.
```
 6   inline void init_bitfields() {
  5     is_channels_last_ = false;
  4     is_channels_last_contiguous_ = false;
  3     is_channels_last_3d_ = false;
  2     is_channels_last_3d_contiguous_ = false;
  1     is_non_overlapping_and_dense_ = true;
```

Reply:
Yeah, this is what I mean by "holistic thinking" here. If we're going to make changes here, all of the fields are going to be involved. It's better to work out what the correct invariants globally are, and then do a PR that gets us into this universe, as opposed to a smaller, "focused" patch that does the bare minimum to fix some problem. The latter approach will result in an endless procession of further bitfixes; if we can establish the correct invariants once and for all, we don't have to fix things.

I haven't thought carefully about how things should be organized. If you want a second opinion here, shout and I'll do the work.